invokeai/app/services/events.py

```diff
@@ -297,3 +290,12 @@ def emit_queue_cleared(self, queue_id: str) -> None:
             event_name="queue_cleared",
             payload=dict(queue_id=queue_id),
         )
+
+    def emit_model_event(self, job: DownloadJobBase):
+        """Emit event when the status of a download/install job changes."""
+        logger = InvokeAILogger.get_logger()
+        progress = 100 * (job.bytes / job.total_bytes) if job.total_bytes > 0 else 0
+        logger.info(f"Dispatch model_event for job {job.id}, status={job.status.value}, progress={progress:5.2f}%")
+        self.dispatch(  # use dispatch() directly here because we are not a session event.
+            event_name="model_event", payload=dict(job=job)
+        )
```

Comment:
How about we set up a separate socket room for model manager events? That should not block merging this PR, would be an immediate followup PR.

Also, I'm not a fan of logging within the event handler - can we move that elsewhere?

Reply:
How do you set up a socket room? I'm not at all familiar with the websockets API. In fact, I'm concerned that there is info in the `DownloadJobBase` object that won't serialize. For example, if the job ends in an error the Exception is placed directly in the object. In the unit test for events, the test subscribes to the in-memory FastAPI exception bus but doesn't check the websocket. Will this be a problem?

Logging is there temporarily for the purpose of debugging events. I'll change the log level to debug or remove it entirely before merging.

Reply:
Yes, the serialization will be a problem. In the events, `error=traceback.format_exc()` provides a nicely formatted error which will display well in the UI.

Note that by the time the payload reaches this place in the code, the error it must be serialized (eg with traceback). This is a requirement from the commercial side of things.

---

To create a new socket room, changes are needed in `invokeai/app/api/sockets.py` and `invokeai/app/services/events.py`. You'd just be copying the existing logic where "queue" is mentioned.

But I don't think we should worry about that in this PR.